kernel/fatal.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <sys/printk.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
+{
+	ARG_UNUSED(reason);
+
+	/* TODO: What's the best way to totally halt the system if SMP
+	 * is enabled?
+	 */
+
+	(void)z_arch_irq_lock();
+	for (;;) {
+		k_cpu_idle();
+	}
+}
+/* LCOV_EXCL_STOP */
+
+/* LCOV_EXCL_START */
+__weak void k_sys_fatal_error_handler(unsigned int reason,
+				      const NANO_ESF *esf)
+{
+	ARG_UNUSED(esf);
+
+	z_fatal_print("Halting system");
+	z_arch_system_halt(reason);
+	CODE_UNREACHABLE;
+}
+/* LCOV_EXCL_STOP */
+
+#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)
+void z_fatal_print(const char *fmt, ...)
+{
+	va_list ap;
+
+	LOG_PANIC();
```

Comment:
This switches logger into panic mode and there is no way back. However he z_fatal_print could is called for recoverable faults.

IMHO, we should call LOG_PANIC() only when we know that system will be halted and the log processing thread will not be called any more.

@nordic-krch: What is your opinion?

Reply:
> This switches logger into panic mode and there is no way back. However he z_fatal_print could is called for recoverable faults.

The problem here is that the policy decision on what to do with the fault doesn't take place until k_sys_fatal_error_handler() runs, and if I understand the logger correctly we need to put the logger in panic mode immediately before dumping any exception info.

Reply:
@pizi-nordic by the way this problem has always existed. I just moved the LOG_PANIC() call to a common spot.

Reply:
@pizi-nordic I filed a bug on this some time ago: #11745

Reply:
@nordic-krch is it safe to move the LOG_PANIC into the k_sys_fatal_error_handler() function? Is there any chance of dropped messages?

Reply:
@andrewboie not calling LOG_PANIC() early will result in logs being deferred but they are not lost. They all will be flushed on `LOG_PANIC()` call and in panic all logs are processed synchronously. However, it's better to call `LOG_PANIC` as early as possible because there is a chance that log message pool is empty and then some messages might be dropped (depending on the mode oldest or newest). So entering panic should be moved to a place where we know that it's unrecoverable.

Reply:
If the configuration of the logger is sane, we could guarantee that last X log messages before call to LOG_PANIC() will be preserved.

The LOG_PANIC() does 2 things:
- Switches backbends to panic mode (so all logs will be processed in place).
- Renders all buffered logs.

Since the call, the LOG() entries are no longer buffered but goes directly to backends.

Reply:
> However, it's better to call LOG_PANIC as early as possible 

> So entering panic should be moved to a place where we know that it's unrecoverable.

This is contradictory. We don't know the exception is unrecoverable until the very end. I think what needs to happen is to provide a mechanism to un-panic the log system.
Pre-existing issue anyway that has its own bug open, don't think it should hold up this PR.

Reply:
right, that sounded a bit contradictory, first part was suppose to be `as early as possible when it's determined that it's unrecoverable`.

I'm against un-panicing. If i assume correctly you may enter fault handling in a thread and scheduler/interrupts are still working and context may be switched. If so then switching to panic mode at that moment may hurt the application since all log calls are now blocking. In the future we may have flash backend which will write/erase from flash, those are time costly operations which when done in blocking way in low priority context will impact other higher priority threads.